packages/xpath/src/functions/xforms/node-set.ts

```diff
@@ -32,6 +34,221 @@ export const countNonEmpty = new NumberFunction(
+	evaluation?: Evaluation
+) => asserts evaluation is LocationPathEvaluation;
+
+/**
+ * @todo This is a concern in several `FunctionImplementation`s. It would be
+ * much nicer if it were handled as part of the signature, then inferred in the
+ * types and validated automatically at runtime. It would also make sense, as a
+ * minor stopgap improvement, to generalize checks like this in a single place
+ * (e.g. as a static method on {@link LocationPathEvaluation} itself). Deferred
+ * here because there is exploratory work on both, but both are out of scope for
+ * work in progress to support {@link indexedRepeat}.
+ */
+const assertIsLocationPathEvaluation: AssertIsLocationPathEvaluation = (evaluation) => {
+	if (!(evaluation instanceof LocationPathEvaluation)) {
+		throw new Error('Expected a node-set result');
+	}
+};
+
+/**
+ * Note: this function is not intended to be general outside of usage by
+ * {@link indexedRepeat}.
+ *
+ * Evaluation of the provided argument is eager—i.e. materializing the complete
+ * array of results, rather than the typical `Iterable<ContextNode>` produced in
+ * most cases—because it is expected that in most cases the eagerness will not
+ * be terribly expensive, and all results will usually be consumed, either to be
+ * indexed or filtered in other ways applicable at call sites.
+ *
+ * Function is named to reflect that expectation.
+ */
+const evaluateArgumentToFilterableNodes = (
+	context: LocationPathEvaluation,
+	arg: EvaluableArgument
+): readonly ContextNode[] => {
+	const evaluation = arg.evaluate(context);
+
+	assertIsLocationPathEvaluation(evaluation);
+
+	return Array.from(evaluation.contextNodes);
+};
+
+interface EvaluatedIndexedRepeatArgumentPair {
+	readonly repeats: readonly ContextNode[];
+	readonly position: number;
+}
+
+type DepthSortResult = -1 | 0 | 1;
+
+/**
+ * @todo This is **obviously cacheable**, but it would make most sense to cache
+ * it at the expression level (or at the expression + bound context node level).
+ * All of the expression analysis machinery is further up the stack (as it
+ * generally ought to be with current designs), but it would be nice to consider
+ * how we'd address caching with these kinds of dynamics at play.
+ */
+const compareContainmentDepth = (
+	{ repeats: a }: EvaluatedIndexedRepeatArgumentPair,
+	{ repeats: b }: EvaluatedIndexedRepeatArgumentPair
+): DepthSortResult => {
+	for (const repeatA of a) {
+		for (const repeatB of b) {
+			if (repeatA.contains(repeatB)) {
+				return -1;
+			}
+
+			if (repeatB.contains(repeatA)) {
+				return 1;
+			}
+		}
+	}
+
+	if (a.length === 0 || b.length === 0) {
+		return 0;
+	}
+
+	// TODO: if we reach this point, there is no hierarchical relationship between
+	// the repeats in `repeatN` and `repeatN + M`. This seems to violate **at
+	// least the intent** of the spec. We should probably produce an error here?
+	return 0;
+};
+
+export const indexedRepeat = new NodeSetFunction(
```

Comment:
Ahh, thanks for flagging this, I wondered about it briefly when reviewing but didn't come back to it. I've never seen attempted usage that results in a nodeset but it certainly could be useful. I'm not sure what Collect will do in that case. I think it would be fine as long as you use it in a context where a nodeset is allowed like `count(indexed-repeat(/data/repeat1/repeat2/foo, /data/repeat1, 3))` so it might be the spec that needs updating. We should try something like that in Collect (I can later if no one beats me to it)

Reply:
I believe I made a case for this in #150. I am open to changing it, but I do want to make the case for it again here.

My interpretation of the spec is that the function...

0. is called in a node-set context (implicit, true of all XPath expressions)
1. produces a new node-set context (intermediate result) for each `repeatN`/`indexN` pair [in containment depth order]
2. resolves to a node-set for the `arg` target (this is the result produced by the spec example's non-shortcut equivalent)
3. ultimately casts that resolved node-set to a string (suggested by spec language: "single node value")

Importantly: XPath semantics only has a concept of node-set value/result type, not a single node. Any single node result—whether returned by a function, or produced by a LocationPath Step—is a node-set with a context size of 1.

In this case, the production of a "single node value" occurs in step 4, where XPath casting semantics produces the string value of the first node in a node-set (regardless of its context size).

The tests clearly demonstrate that behavior: anywhere a string value is consumed, the expected string value is produced from the single (first) node value in the resulting node-set.

In `@getodk/xpath` this casting from node-set to string occurs further up the call stack, applied by intentionally isolated casting logic, and applied consistently:

- to any XPath result
- of any type
- whether an intermediate result of a sub-expression, or the ultimate result produced by an `evaluate` call
- whether the result is produced by a `FunctionImplementation`, any part of a LocationPath, or any other result from any other XPath syntax

Our isolation of casting logic is important. XPath casting semantics have a fair bit of nuance which can easily be lost or become inconsistent otherwise (and did earlier in implementation!). That isolation is also crucial to our ability to support some ODK XForms functionality, i.e. date / time / datetime data types.

- - -

In terms of spec conformance: the only way to deviate from the spec behavior is to intentionally consume the return value **as a node-set**, as in @lognaturel's `count` example, or by calling `evaluate` with any of the `XPathResult` types producing one or more nodes **and then explicitly accessing the evaluated result's node APIs, or by calling any of our explicitly node-returning `evaluate*` convenience methods; any of which are explicit and intentional. Doing so does go **beyond the spec**, but it does not conflict with the spec. This is the same reasoning I applied to implementing the function with a variadic signature (rather than restricting calls to a maximum of three `repeatN`/`indexN` pairs as in the spec).

Since the spec signature returns a string, there's no precedent for _different behavior_ when an `indexed-repeat` return value is deliberately used as a node-set: you can't cast a string _to a node-set_.

The only conflict this interpretation could introduce with the spec is if we think it's more valuable to users to actively prevent the beyond-spec usage than to (quietly, apart from this discussion) permit it.

Reply:
Thanks for the explanation, it feels fine to leave as-is to me. 👍 @sadiqkhoja?

Reply:
I just saw the test below and the edited comment. Ok, seems like there's more to consider!